engine/src/main/java/org/pentaho/di/trans/steps/httppost/HTTPPOST.java

```diff
@@ -0,0 +1,558 @@
+              BaseMessages.getString( PKG, "HTTPPOST.Log.ErrorFindingField" ) + meta.getQueryField()[ i ] + "]" );
+            throw new KettleStepException( BaseMessages.getString( PKG, "HTTPPOST.Exception.CouldnotFindField", meta
+              .getQueryField()[ i ] ) );
+          }
+          data.queryParameters[ i ] =
+            new BasicNameValuePair( environmentSubstitute( meta.getQueryParameter()[ i ] ),
+              data.outputRowMeta.getString( r,
+                data.query_parameters_nrs[ i ] ) );
+        }
+      }
+      // set request entity?
+      if ( !Utils.isEmpty( meta.getRequestEntity() ) ) {
+        data.indexOfRequestEntity = data.inputRowMeta.indexOfValue( environmentSubstitute( meta.getRequestEntity() ) );
+        if ( data.indexOfRequestEntity < 0 ) {
+          throw new KettleStepException( BaseMessages.getString( PKG,
+            "HTTPPOST.Exception.CouldnotFindRequestEntityField", meta.getRequestEntity() ) );
+        }
+      }
+      data.realEncoding = environmentSubstitute( meta.getEncoding() );
+    } // end if first
+
+    try {
+      Object[] outputRowData = callHTTPPOST( r );
+      putRow( data.outputRowMeta, outputRowData ); // copy row to output rowset(s);
+
+      if ( checkFeedback( getLinesRead() ) ) {
+        if ( isDetailed() ) {
+          logDetailed( BaseMessages.getString( PKG, "HTTPPOST.LineNumber" ) + getLinesRead() );
+        }
+      }
+    } catch ( KettleException e ) {
+      boolean sendToErrorRow = false;
+      String errorMessage = null;
+
+      if ( getStepMeta().isDoingErrorHandling() ) {
+        sendToErrorRow = true;
+        errorMessage = e.toString();
+      } else {
+        logError( BaseMessages.getString( PKG, "HTTPPOST.ErrorInStepRunning" ) + e.getMessage() );
+        setErrors( 1 );
+        logError( Const.getStackTracker( e ) );
+        stopAll();
+        setOutputDone(); // signal end to receiver(s)
+        return false;
+      }
+
+      if ( sendToErrorRow ) {
+        // Simply add this row to the error row
+        putError( getInputRowMeta(), r, 1, errorMessage, null, "HTTPPOST001" );
+      }
+
+    }
+
+    return true;
+  }
+
+  @VisibleForTesting
+  String getRequestBodyParamsAsStr( NameValuePair[] pairs, String charset ) throws KettleException {
+    StringBuffer buf = new StringBuffer();
+    try {
+      for ( int i = 0; i < pairs.length; ++i ) {
+        NameValuePair pair = pairs[ i ];
+        if ( pair.getName() != null ) {
+          if ( i > 0 ) {
+            buf.append( "&" );
+          }
+
+          buf.append( URLEncoder.encode( pair.getName(), !StringUtil.isEmpty( charset ) ? charset : DEFAULT_ENCODING ) );
+          buf.append( "=" );
+          if ( pair.getValue() != null ) {
+            buf.append( URLEncoder.encode( pair.getValue(), !StringUtil.isEmpty( charset ) ? charset : DEFAULT_ENCODING ) );
+          }
+        }
+      }
+      return buf.toString();
+    } catch ( UnsupportedEncodingException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)